--- lz4-v1.9.4/lib/lz4.c	2022-08-16 06:45:30
+++ lz4-opt/lib/lz4.c	2026-05-29 09:20:28
@@ -55,8 +55,68 @@
  * get treated as LZ4_ACCELERATION_MAX instead (fix #876)
  */
 #define LZ4_ACCELERATION_MAX 65537
+
+#include <arm_neon.h>
+#include <stddef.h>
+
+#define G_64KLIMIT       ((65536) + (11))
+
+#define G_KZLPRIME5BYTES (889523592379ULL)
+#define G_KZLPRIME8BYTES (11400714785074694791ULL)
+
+#ifndef KZL_FORCE_INLINE
+#  ifdef _MSC_VER
+#    define KZL_FORCE_INLINE static __forceinline
+#  else
+#    if defined (__cplusplus) || defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L   /* C99 */
+#      ifdef __GNUC__
+#        define KZL_FORCE_INLINE static inline __attribute__((always_inline))
+#      else
+#        define KZL_FORCE_INLINE static inline
+#      endif
+#    else
+#      define KZL_FORCE_INLINE static
+#    endif
+#  endif
+#endif
+
+#if defined(__GNUC__) && (__GNUC__ >= 4)
+#    define KZL_MEMCPY_2(dst, src, size) __builtin_memcpy(dst, src, size)
+#    define KZL_MEMCPY_4(dst, src, size) __builtin_memcpy(dst, src, size)
+#    define KZL_MEMCPY_8(dst, src, size) vst1_u8((dst), vld1_u8(src))
+#    define KZL_MEMCPY_16(dst, src, size) vst1q_u8((dst), vld1q_u8(src))
+#    define KZL_MEMCPY_32(dst, src, size) vst1q_u8((dst), vld1q_u8(src)); vst1q_u8(((dst)+16), vld1q_u8(((src)+16)))
+#    define KZL_MEMCPY_16X1(dst, src, size) vst1q_u64((dst), vld1q_u64(src))
+#    define KZL_MEMCPY_32X1(dst, src, size) vst1q_u64((dst), vld1q_u64(src)); \
+                vst1q_u64(((dst)+16), vld1q_u64(((src)+16)))
+#endif
+
+#define SEQUENCE_MOVE 24
+#define HASH_SIZE 64
 
+KZL_FORCE_INLINE uint32_t KZL_LittleEndianfastHash5(uint64_t sequence, uint8_t HashLogUsage)
+{
+    return (uint32_t)(((sequence << SEQUENCE_MOVE) * G_KZLPRIME5BYTES) >> (HASH_SIZE - HashLogUsage));
+}
 
+KZL_FORCE_INLINE uint32_t KZL_BigEndianfastHash5(uint64_t sequence, uint8_t HashLogUsage)
+{
+    return (uint32_t)(((sequence >> SEQUENCE_MOVE) * G_KZLPRIME8BYTES) >> (HASH_SIZE - HashLogUsage));
+}
+
+static void skipTrigger(int srcSize, uint8_t *skipStep)
+{
+    if (skipStep == NULL || srcSize <= 0) {
+        return;
+    }
+
+    if (srcSize >= G_64KLIMIT) {
+        *skipStep = 2;
+    } else {
+        *skipStep = 4;
+    }
+}
+
 /*-************************************
 *  CPU Feature Detection
 **************************************/
@@ -272,7 +332,6 @@ static const int LZ4_minLength = (MFLIMIT+1);
 #  endif
 #endif
 
-#define LZ4_STATIC_ASSERT(c)   { enum { LZ4_static_assert = 1/(int)(!!(c)) }; }   /* use after variable declarations */
 
 #if defined(LZ4_DEBUG) && (LZ4_DEBUG>=2)
 #  include <stdio.h>
@@ -283,8 +342,10 @@ static const int LZ4_minLength = (MFLIMIT+1);
             fprintf(stderr, __VA_ARGS__);             \
             fprintf(stderr, " \n");                   \
     }   }
+#  define LZ4_STATIC_ASSERT(c)   { enum { LZ4_static_assert = 1/(int)(!!(c)) }; }   /* use after variable declarations */
 #else
 #  define DEBUGLOG(l, ...) {}    /* disabled */
+#  define LZ4_STATIC_ASSERT(c)   {}   /* disabled */
 #endif
 
 static int LZ4_isAligned(const void* ptr, size_t alignment)
@@ -447,7 +508,17 @@ void LZ4_wildCopy8(void* dstPtr, const void* srcPtr, v
     const BYTE* s = (const BYTE*)srcPtr;
     BYTE* const e = (BYTE*)dstEnd;
 
-    do { LZ4_memcpy(d,s,8); d+=8; s+=8; } while (d<e);
+    do { KZL_MEMCPY_8(d,s,8); d+=8; s+=8; } while (d<e);
+}
+
+LZ4_FORCE_INLINE
+void LZ4_wildCopy16(void* dstPtr, const void* srcPtr, void* dstEnd)
+{
+    BYTE* d = (BYTE*)dstPtr;
+    const BYTE* s = (const BYTE*)srcPtr;
+    BYTE* const e = (BYTE*)dstEnd;
+
+    do { LZ4_memcpy(d,s,16); d+=16; s+=16; } while (d<e);
 }
 
 static const unsigned inc32table[8] = {0, 1, 2,  1,  0,  4, 4, 4};
@@ -482,11 +553,11 @@ LZ4_memcpy_using_offset_base(BYTE* dstPtr, const BYTE*
         dstPtr[2] = srcPtr[2];
         dstPtr[3] = srcPtr[3];
         srcPtr += inc32table[offset];
-        LZ4_memcpy(dstPtr+4, srcPtr, 4);
+        KZL_MEMCPY_4(dstPtr+4, srcPtr, 4);
         srcPtr -= dec64table[offset];
         dstPtr += 8;
     } else {
-        LZ4_memcpy(dstPtr, srcPtr, 8);
+        KZL_MEMCPY_8(dstPtr, srcPtr, 8);
         dstPtr += 8;
         srcPtr += 8;
     }
@@ -504,9 +575,27 @@ LZ4_wildCopy32(void* dstPtr, const void* srcPtr, void*
     const BYTE* s = (const BYTE*)srcPtr;
     BYTE* const e = (BYTE*)dstEnd;
 
-    do { LZ4_memcpy(d,s,16); LZ4_memcpy(d+16,s+16,16); d+=32; s+=32; } while (d<e);
+    do { KZL_MEMCPY_32(d,s,32); d+=32; s+=32; __builtin_prefetch(s,0,0);} while (d<e);
 }
 
+LZ4_FORCE_INLINE void
+LZ4_wildCopy64(void* dstPtr, const void* srcPtr, void* dstEnd)
+{
+    BYTE* d = (BYTE*)dstPtr;
+    const BYTE* s = (const BYTE*)srcPtr;
+    BYTE* const e = (BYTE*)dstEnd;
+
+    do {
+        LZ4_memcpy(d,s,16);
+        LZ4_memcpy(d+16,s+16,16);
+        LZ4_memcpy(d+32,s+32,16);
+        LZ4_memcpy(d+48,s+48,16);
+
+        d+=64; s+=64;
+    } while (d<e);
+}
+
+
 /* LZ4_memcpy_using_offset()  presumes :
  * - dstEnd >= dstPtr + MINMATCH
  * - there is at least 8 bytes available to write after dstEnd */
@@ -522,30 +611,30 @@ LZ4_memcpy_using_offset(BYTE* dstPtr, const BYTE* srcP
         MEM_INIT(v, *srcPtr, 8);
         break;
     case 2:
-        LZ4_memcpy(v, srcPtr, 2);
-        LZ4_memcpy(&v[2], srcPtr, 2);
+        KZL_MEMCPY_2(v, srcPtr, 2);
+        KZL_MEMCPY_2(v+2, srcPtr, 2);
 #if defined(_MSC_VER) && (_MSC_VER <= 1933) /* MSVC 2022 ver 17.3 or earlier */
 #  pragma warning(push)
 #  pragma warning(disable : 6385) /* warning C6385: Reading invalid data from 'v'. */
 #endif
-        LZ4_memcpy(&v[4], v, 4);
+        KZL_MEMCPY_4(v+4, v, 4);
 #if defined(_MSC_VER) && (_MSC_VER <= 1933) /* MSVC 2022 ver 17.3 or earlier */
 #  pragma warning(pop)
 #endif
         break;
     case 4:
-        LZ4_memcpy(v, srcPtr, 4);
-        LZ4_memcpy(&v[4], srcPtr, 4);
+        KZL_MEMCPY_4(v, srcPtr, 4);
+        KZL_MEMCPY_4(v+4, srcPtr, 4);
         break;
     default:
         LZ4_memcpy_using_offset_base(dstPtr, srcPtr, dstEnd, offset);
         return;
     }
 
-    LZ4_memcpy(dstPtr, v, 8);
+    KZL_MEMCPY_8(dstPtr, v, 8);
     dstPtr += 8;
     while (dstPtr < dstEnd) {
-        LZ4_memcpy(dstPtr, v, 8);
+        KZL_MEMCPY_8(dstPtr, v, 8);
         dstPtr += 8;
     }
 }
@@ -753,6 +842,8 @@ int LZ4_decompress_safe_partial_forceExtDict(const cha
 /*-******************************
 *  Compression functions
 ********************************/
+#define g_prime5bytes 889523592379ULL
+
 LZ4_FORCE_INLINE U32 LZ4_hash4(U32 sequence, tableType_t const tableType)
 {
     if (tableType == byU16)
@@ -764,12 +855,10 @@ LZ4_FORCE_INLINE U32 LZ4_hash5(U64 sequence, tableType
 LZ4_FORCE_INLINE U32 LZ4_hash5(U64 sequence, tableType_t const tableType)
 {
     const U32 hashLog = (tableType == byU16) ? LZ4_HASHLOG+1 : LZ4_HASHLOG;
-    if (LZ4_isLittleEndian()) {
-        const U64 prime5bytes = 889523592379ULL;
-        return (U32)(((sequence << 24) * prime5bytes) >> (64 - hashLog));
+    if (__builtin_expect(LZ4_isLittleEndian(), 1)) {
+        return KZL_LittleEndianfastHash5(sequence, hashLog);
     } else {
-        const U64 prime8bytes = 11400714785074694791ULL;
-        return (U32)(((sequence >> 24) * prime8bytes) >> (64 - hashLog));
+        return KZL_BigEndianfastHash5(sequence, hashLog);
     }
 }
 
@@ -777,6 +866,11 @@ LZ4_FORCE_INLINE U32 LZ4_hashPosition(const void* cons
 {
     if ((sizeof(reg_t)==8) && (tableType != byU16)) return LZ4_hash5(LZ4_read_ARCH(p), tableType);
     return LZ4_hash4(LZ4_read32(p), tableType);
+}
+
+LZ4_FORCE_INLINE U32 LZ4_hashPositionIgnoreBufferLength(const void* const p, tableType_t const tableType)
+{
+    return LZ4_hash5(LZ4_read_ARCH(p), tableType);
 }
 
 LZ4_FORCE_INLINE void LZ4_clearHash(U32 h, void* tableBase, tableType_t const tableType)
@@ -785,9 +879,9 @@ LZ4_FORCE_INLINE void LZ4_clearHash(U32 h, void* table
     {
     default: /* fallthrough */
     case clearedTable: { /* illegal! */ assert(0); return; }
-    case byPtr: { const BYTE** hashTable = (const BYTE**)tableBase; hashTable[h] = NULL; return; }
     case byU32: { U32* hashTable = (U32*) tableBase; hashTable[h] = 0; return; }
     case byU16: { U16* hashTable = (U16*) tableBase; hashTable[h] = 0; return; }
+    case byPtr: { const BYTE** hashTable = (const BYTE**)tableBase; hashTable[h] = NULL; return; }
     }
 }
 
@@ -797,9 +891,9 @@ LZ4_FORCE_INLINE void LZ4_putIndexOnHash(U32 idx, U32 
     {
     default: /* fallthrough */
     case clearedTable: /* fallthrough */
-    case byPtr: { /* illegal! */ assert(0); return; }
     case byU32: { U32* hashTable = (U32*) tableBase; hashTable[h] = idx; return; }
     case byU16: { U16* hashTable = (U16*) tableBase; assert(idx < 65536); hashTable[h] = (U16)idx; return; }
+    case byPtr: { /* illegal! */ assert(0); return; }
     }
 }
 
@@ -810,15 +904,15 @@ LZ4_FORCE_INLINE void LZ4_putPositionOnHash(const BYTE
     switch (tableType)
     {
     case clearedTable: { /* illegal! */ assert(0); return; }
-    case byPtr: { const BYTE** hashTable = (const BYTE**)tableBase; hashTable[h] = p; return; }
     case byU32: { U32* hashTable = (U32*) tableBase; hashTable[h] = (U32)(p-srcBase); return; }
     case byU16: { U16* hashTable = (U16*) tableBase; hashTable[h] = (U16)(p-srcBase); return; }
+    case byPtr: { const BYTE** hashTable = (const BYTE**)tableBase; hashTable[h] = p; return; }
     }
 }
 
 LZ4_FORCE_INLINE void LZ4_putPosition(const BYTE* p, void* tableBase, tableType_t tableType, const BYTE* srcBase)
 {
-    U32 const h = LZ4_hashPosition(p, tableType);
+    U32 const h = LZ4_hashPositionIgnoreBufferLength(p, tableType);
     LZ4_putPositionOnHash(p, h, tableBase, tableType, srcBase);
 }
 
@@ -846,7 +940,7 @@ static const BYTE* LZ4_getPositionOnHash(U32 h, const 
 
 static const BYTE* LZ4_getPositionOnHash(U32 h, const void* tableBase, tableType_t tableType, const BYTE* srcBase)
 {
-    if (tableType == byPtr) { const BYTE* const* hashTable = (const BYTE* const*) tableBase; return hashTable[h]; }
+    if (unlikely(tableType == byPtr)) { const BYTE* const* hashTable = (const BYTE* const*) tableBase; return hashTable[h]; }
     if (tableType == byU32) { const U32* const hashTable = (const U32*) tableBase; return hashTable[h] + srcBase; }
     { const U16* const hashTable = (const U16*) tableBase; return hashTable[h] + srcBase; }   /* default, to ensure a return */
 }
@@ -856,7 +950,7 @@ LZ4_getPosition(const BYTE* p,
                 const void* tableBase, tableType_t tableType,
                 const BYTE* srcBase)
 {
-    U32 const h = LZ4_hashPosition(p, tableType);
+    U32 const h = LZ4_hashPositionIgnoreBufferLength(p, tableType);
     return LZ4_getPositionOnHash(h, tableBase, tableType, srcBase);
 }
 
@@ -870,11 +964,10 @@ LZ4_prepareTable(LZ4_stream_t_internal* const cctx,
      */
     if ((tableType_t)cctx->tableType != clearedTable) {
         assert(inputSize >= 0);
-        if ((tableType_t)cctx->tableType != tableType
+        if (inputSize >= 4 KB || (tableType_t)cctx->tableType != tableType
           || ((tableType == byU16) && cctx->currentOffset + (unsigned)inputSize >= 0xFFFFU)
           || ((tableType == byU32) && cctx->currentOffset > 1 GB)
-          || tableType == byPtr
-          || inputSize >= 4 KB)
+          || tableType == byPtr)
         {
             DEBUGLOG(4, "LZ4_prepareTable: Resetting table in %p", cctx);
             MEM_INIT(cctx->hashTable, 0, LZ4_HASHTABLESIZE);
@@ -967,7 +1060,7 @@ LZ4_FORCE_INLINE int LZ4_compress_generic_validated(
     lowLimit = (const BYTE*)source - (dictDirective == withPrefix64k ? dictSize : 0);
 
     /* Update context state */
-    if (dictDirective == usingDictCtx) {
+    if (unlikely(dictDirective == usingDictCtx)) {
         /* Subsequent linked blocks can't use the dictionary. */
         /* Instead, they use the block we just compressed. */
         cctx->dictCtx = NULL;
@@ -982,30 +1075,33 @@ LZ4_FORCE_INLINE int LZ4_compress_generic_validated(
 
     /* First Byte */
     LZ4_putPosition(ip, cctx->hashTable, tableType, base);
-    ip++; forwardH = LZ4_hashPosition(ip, tableType);
+    ip++; forwardH = LZ4_hashPositionIgnoreBufferLength(ip, tableType);
+    uint8_t skipStep = LZ4_skipTrigger;
+    skipTrigger(inputSize, &skipStep);
 
     /* Main Loop */
+    __asm__(".p2align 6");
     for ( ; ; ) {
         const BYTE* match;
         BYTE* token;
         const BYTE* filledIp;
 
         /* Find a match */
-        if (tableType == byPtr) {
+        if (unlikely(tableType == byPtr)) {
             const BYTE* forwardIp = ip;
             int step = 1;
-            int searchMatchNb = acceleration << LZ4_skipTrigger;
+            int searchMatchNb = acceleration << skipStep;
             do {
                 U32 const h = forwardH;
                 ip = forwardIp;
                 forwardIp += step;
-                step = (searchMatchNb++ >> LZ4_skipTrigger);
+                step = (searchMatchNb++ >> skipStep);
 
                 if (unlikely(forwardIp > mflimitPlusOne)) goto _last_literals;
                 assert(ip < mflimitPlusOne);
 
                 match = LZ4_getPositionOnHash(h, cctx->hashTable, tableType, base);
-                forwardH = LZ4_hashPosition(forwardIp, tableType);
+                forwardH = LZ4_hashPositionIgnoreBufferLength(forwardIp, tableType);
                 LZ4_putPositionOnHash(ip, h, cctx->hashTable, tableType, base);
 
             } while ( (match+LZ4_DISTANCE_MAX < ip)
@@ -1015,7 +1111,8 @@ LZ4_FORCE_INLINE int LZ4_compress_generic_validated(
 
             const BYTE* forwardIp = ip;
             int step = 1;
-            int searchMatchNb = acceleration << LZ4_skipTrigger;
+            int searchMatchNb = acceleration << skipStep;
+            __asm__(".p2align 6");
             do {
                 U32 const h = forwardH;
                 U32 const current = (U32)(forwardIp - base);
@@ -1024,12 +1121,12 @@ LZ4_FORCE_INLINE int LZ4_compress_generic_validated(
                 assert(forwardIp - base < (ptrdiff_t)(2 GB - 1));
                 ip = forwardIp;
                 forwardIp += step;
-                step = (searchMatchNb++ >> LZ4_skipTrigger);
+                step = (searchMatchNb++ >> skipStep);
 
                 if (unlikely(forwardIp > mflimitPlusOne)) goto _last_literals;
                 assert(ip < mflimitPlusOne);
 
-                if (dictDirective == usingDictCtx) {
+                if (unlikely(dictDirective == usingDictCtx)) {
                     if (matchIndex < startIndex) {
                         /* there was no match, try the dictionary */
                         assert(tableType == byU32);
@@ -1041,7 +1138,7 @@ LZ4_FORCE_INLINE int LZ4_compress_generic_validated(
                         match = base + matchIndex;
                         lowLimit = (const BYTE*)source;
                     }
-                } else if (dictDirective == usingExtDict) {
+                } else if (unlikely(dictDirective == usingExtDict)) {
                     if (matchIndex < startIndex) {
                         DEBUGLOG(7, "extDict candidate: matchIndex=%5u  <  startIndex=%5u", matchIndex, startIndex);
                         assert(startIndex - matchIndex >= MINMATCH);
@@ -1055,11 +1152,11 @@ LZ4_FORCE_INLINE int LZ4_compress_generic_validated(
                 } else {   /* single continuous memory segment */
                     match = base + matchIndex;
                 }
-                forwardH = LZ4_hashPosition(forwardIp, tableType);
+                forwardH = LZ4_hashPositionIgnoreBufferLength(forwardIp, tableType);
                 LZ4_putIndexOnHash(current, h, cctx->hashTable, tableType);
 
                 DEBUGLOG(7, "candidate at pos=%u  (offset=%u \n", matchIndex, current - matchIndex);
-                if ((dictIssue == dictSmall) && (matchIndex < prefixIdxLimit)) { continue; }    /* match outside of valid area */
+                if (unlikely(dictIssue == dictSmall) && (matchIndex < prefixIdxLimit)) { continue; }    /* match outside of valid area */
                 assert(matchIndex < current);
                 if ( ((tableType != byU16) || (LZ4_DISTANCE_MAX < LZ4_DISTANCE_ABSOLUTE_MAX))
                   && (matchIndex+LZ4_DISTANCE_MAX < current)) {
@@ -1077,22 +1174,34 @@ LZ4_FORCE_INLINE int LZ4_compress_generic_validated(
 
         /* Catch up */
         filledIp = ip;
+        assert(ip > anchor); /* this is always true as ip has been advanced before entering the main loop */
         while (((ip>anchor) & (match > lowLimit)) && (unlikely(ip[-1]==match[-1]))) { ip--; match--; }
 
         /* Encode Literals */
         {   unsigned const litLength = (unsigned)(ip - anchor);
             token = op++;
             if ((outputDirective == limitedOutput) &&  /* Check output buffer overflow */
-                (unlikely(op + litLength + (2 + 1 + LASTLITERALS) + (litLength/255) > olimit)) ) {
-                return 0;   /* cannot compress within `dst` budget. Stored indexes in hash table are nonetheless fine */
+                 (unlikely(op + litLength + (2 + 1 + LASTLITERALS) + (litLength/255) + 8 /* To catch cases where `LZ4_wildCopy16` fails */ > olimit)) ) {
+                 if(op + litLength + (2 + 1 + LASTLITERALS) + (litLength/255)> olimit)
+                     return 0;
+                 if (litLength >= RUN_MASK) {
+                     unsigned len = litLength - RUN_MASK;
+                     *token = (RUN_MASK<<ML_BITS);
+                     for(; len >= 255 ; len-=255) *op++ = 255;
+                     *op++ = (BYTE)len;
+                 }
+                 else *token = (BYTE)(litLength<<ML_BITS);
+                 LZ4_wildCopy8(op, anchor, op+litLength);
+                 goto _skip_16_bytes_copy;
             }
-            if ((outputDirective == fillOutput) &&
-                (unlikely(op + (litLength+240)/255 /* litlen */ + litLength /* literals */ + 2 /* offset */ + 1 /* token */ + MFLIMIT - MINMATCH /* min last literals so last match is <= end - MFLIMIT */ > olimit))) {
+
+            if ((outputDirective == fillOutput) && /* To catch cases where `LZ4_wildCopy16` fails */
+                (unlikely(op + (litLength+240)/255 /* litlen */ + litLength /* literals */ + 2 /* offset */ + 1 /* token */ + MFLIMIT - MINMATCH + 8 /* min last literals so last match is <= end - MFLIMIT */ > olimit))) {
                 op--;
                 goto _last_literals;
             }
             if (litLength >= RUN_MASK) {
-                int len = (int)(litLength - RUN_MASK);
+                unsigned len = litLength - RUN_MASK;
                 *token = (RUN_MASK<<ML_BITS);
                 for(; len >= 255 ; len-=255) *op++ = 255;
                 *op++ = (BYTE)len;
@@ -1100,7 +1209,8 @@ LZ4_FORCE_INLINE int LZ4_compress_generic_validated(
             else *token = (BYTE)(litLength<<ML_BITS);
 
             /* Copy Literals */
-            LZ4_wildCopy8(op, anchor, op+litLength);
+            LZ4_wildCopy16(op, anchor, op+litLength);
+_skip_16_bytes_copy:
             op+=litLength;
             DEBUGLOG(6, "seq.start:%i, literals=%u, match.start:%i",
                         (int)(anchor-(const BYTE*)source), litLength, (int)(ip-(const BYTE*)source));
@@ -1115,7 +1225,7 @@ _next_match:
          * - token and *token : position to write 4-bits for match length; higher 4-bits for literal length supposed already written
          */
 
-        if ((outputDirective == fillOutput) &&
+        if (unlikely(outputDirective == fillOutput) &&
             (op + 2 /* offset */ + 1 /* token */ + MFLIMIT - MINMATCH /* min last literals so last match is <= end - MFLIMIT */ > olimit)) {
             /* the match was too close to the end, rewind and go to last literals */
             op = token;
@@ -1136,7 +1246,7 @@ _next_match:
         /* Encode MatchLength */
         {   unsigned matchCode;
 
-            if ( (dictDirective==usingExtDict || dictDirective==usingDictCtx)
+            if ( unlikely(dictDirective==usingExtDict || dictDirective==usingDictCtx)
               && (lowLimit==dictionary) /* match within extDict */ ) {
                 const BYTE* limit = ip + (dictEnd-match);
                 assert(dictEnd > match);
@@ -1155,7 +1265,7 @@ _next_match:
                 DEBUGLOG(6, "             with matchLength=%u", matchCode+MINMATCH);
             }
 
-            if ((outputDirective) &&    /* Check output buffer overflow */
+            if (unlikely(outputDirective) &&    /* Check output buffer overflow */
                 (unlikely(op + (1 + LASTLITERALS) + (matchCode+240)/255 > olimit)) ) {
                 if (outputDirective == fillOutput) {
                     /* Match description too long : reduce it */
@@ -1172,7 +1282,7 @@ _next_match:
                         const BYTE* ptr;
                         DEBUGLOG(5, "Clearing %u positions", (U32)(filledIp - ip));
                         for (ptr = ip; ptr <= filledIp; ++ptr) {
-                            U32 const h = LZ4_hashPosition(ptr, tableType);
+                            U32 const h = LZ4_hashPositionIgnoreBufferLength(ptr, tableType);
                             LZ4_clearHash(h, cctx->hashTable, tableType);
                         }
                     }
@@ -1217,11 +1327,11 @@ _next_match:
 
         } else {   /* byU32, byU16 */
 
-            U32 const h = LZ4_hashPosition(ip, tableType);
+            U32 const h = LZ4_hashPositionIgnoreBufferLength(ip, tableType);
             U32 const current = (U32)(ip-base);
             U32 matchIndex = LZ4_getIndexOnHash(h, cctx->hashTable, tableType);
             assert(matchIndex < current);
-            if (dictDirective == usingDictCtx) {
+            if (unlikely(dictDirective == usingDictCtx)) {
                 if (matchIndex < startIndex) {
                     /* there was no match, try the dictionary */
                     matchIndex = LZ4_getIndexOnHash(h, dictCtx->hashTable, byU32);
@@ -1232,7 +1342,7 @@ _next_match:
                     match = base + matchIndex;
                     lowLimit = (const BYTE*)source;  /* required for match length counter */
                 }
-            } else if (dictDirective==usingExtDict) {
+            } else if (unlikely(dictDirective==usingExtDict)) {
                 if (matchIndex < startIndex) {
                     assert(dictBase);
                     match = dictBase + matchIndex;
@@ -1259,7 +1369,7 @@ _next_match:
         }
 
         /* Prepare next loop */
-        forwardH = LZ4_hashPosition(++ip, tableType);
+        forwardH = LZ4_hashPositionIgnoreBufferLength(++ip, tableType);
 
     }
 
@@ -1322,7 +1432,7 @@ LZ4_FORCE_INLINE int LZ4_compress_generic(
                 srcSize, dstCapacity);
 
     if ((U32)srcSize > (U32)LZ4_MAX_INPUT_SIZE) { return 0; }  /* Unsupported srcSize, too large (or negative) */
-    if (srcSize == 0) {   /* src == NULL supported if srcSize == 0 */
+    if (unlikely(srcSize == 0)) {   /* src == NULL supported if srcSize == 0 */
         if (outputDirective != notLimited && dstCapacity <= 0) return 0;  /* no output, can't write anything */
         DEBUGLOG(5, "Generating an empty block");
         assert(outputDirective == notLimited || dstCapacity >= 1);
@@ -1908,7 +2018,7 @@ read_variable_length(const BYTE** ip, const BYTE* ilim
     assert(ip != NULL);
     assert(*ip !=  NULL);
     assert(ilimit != NULL);
-    if (initial_check && unlikely((*ip) >= ilimit)) {    /* read limit reached */
+    if (unlikely((*ip) >= ilimit) && initial_check) {    /* read limit reached */
         return rvl_error;
     }
     do {
@@ -1919,7 +2029,7 @@ read_variable_length(const BYTE** ip, const BYTE* ilim
             return rvl_error;
         }
         /* accumulator overflow detection (32-bit mode only) */
-        if ((sizeof(length)<8) && unlikely(length > ((Rvl_t)(-1)/2)) ) {
+        if (unlikely(sizeof(length)<8) && unlikely(length > ((Rvl_t)(-1)/2)) ) {
             return rvl_error;
         }
     } while (s==255);
@@ -1987,12 +2097,13 @@ LZ4_decompress_generic(
      * where copying reliably 32-bytes is preferable to an unpredictable branch.
      * note : fast loop may show a regression for some client arm chips. */
 #if LZ4_FAST_DEC_LOOP
-        if ((oend - op) < FASTLOOP_SAFE_DISTANCE) {
+        if (unlikely((oend - op) < FASTLOOP_SAFE_DISTANCE)) {
             DEBUGLOG(6, "skip fast decode loop");
             goto safe_decode;
         }
 
         /* Fast loop : decode sequences as long as output < oend-FASTLOOP_SAFE_DISTANCE */
+        __asm__(".p2align 6");
         while (1) {
             /* Main fastloop assertion: We can always wildcopy FASTLOOP_SAFE_DISTANCE */
             assert(oend - op >= FASTLOOP_SAFE_DISTANCE);
@@ -2001,7 +2112,8 @@ LZ4_decompress_generic(
             length = token >> ML_BITS;  /* literal length */
 
             /* decode literal length */
-            if (length == RUN_MASK) {
+            if(unlikely(token >= (RUN_MASK << ML_BITS))) {
+                __builtin_prefetch(op + 64, 0, 3);
                 size_t const addl = read_variable_length(&ip, iend-RUN_MASK, 1);
                 if (addl == rvl_error) { goto _output_error; }
                 length += addl;
@@ -2009,10 +2121,11 @@ LZ4_decompress_generic(
                 if (unlikely((uptrval)(ip)+length<(uptrval)(ip))) { goto _output_error; } /* overflow detection */
 
                 /* copy literals */
+                __builtin_prefetch(ip, 0, 1);
                 cpy = op+length;
                 LZ4_STATIC_ASSERT(MFLIMIT >= WILDCOPYLENGTH);
-                if ((cpy>oend-32) || (ip+length>iend-32)) { goto safe_literal_copy; }
-                LZ4_wildCopy32(op, ip, cpy);
+                if ((op+length>oend-FASTLOOP_SAFE_DISTANCE) || (ip+length>iend-FASTLOOP_SAFE_DISTANCE)) { goto safe_literal_copy; }
+                LZ4_wildCopy64(op, ip, op+length);
                 ip += length; op = cpy;
             } else {
                 cpy = op+length;
@@ -2020,7 +2133,7 @@ LZ4_decompress_generic(
                 /* We don't need to check oend, since we check it once for each loop below */
                 if (ip > iend-(16 + 1/*max lit + offset + nextToken*/)) { goto safe_literal_copy; }
                 /* Literals can only be <= 14, but hope compilers optimize better when copy by a register size */
-                LZ4_memcpy(op, ip, 16);
+                KZL_MEMCPY_16(op, ip, 16);
                 ip += length; op = cpy;
             }
 
@@ -2049,25 +2162,34 @@ LZ4_decompress_generic(
                 }
 
                 /* Fastpath check: skip LZ4_wildCopy32 when true */
-                if ((dict == withPrefix64k) || (match >= lowPrefix)) {
+                if (unlikely((dict == withPrefix64k) || (match >= lowPrefix))) {
                     if (offset >= 8) {
+                        const BYTE* match_1 = match;
+                        const BYTE* match_2 = match+8;
+                        const BYTE* match_3 = match+16;
+                        __builtin_prefetch(match_1, 0, 1);
+                        __builtin_prefetch(match_2, 0, 1);
+                        __builtin_prefetch(match_3, 0, 1);
+
                         assert(match >= lowPrefix);
                         assert(match <= op);
                         assert(op + 18 <= oend);
 
-                        LZ4_memcpy(op, match, 8);
-                        LZ4_memcpy(op+8, match+8, 8);
-                        LZ4_memcpy(op+16, match+16, 2);
+                        KZL_MEMCPY_8(op, match_1, 8);
+                        KZL_MEMCPY_8(op+8, match_2, 8);
+                        KZL_MEMCPY_2(op+16, match_3, 2);
                         op += length;
                         continue;
-            }   }   }
+                    }   
+                }   
+            }
 
             if (checkOffset && (unlikely(match + dictSize < lowPrefix))) { goto _output_error; } /* Error : offset outside buffers */
             /* match starting within external dictionary */
-            if ((dict==usingExtDict) && (match < lowPrefix)) {
+            if (unlikely((dict==usingExtDict) && (match < lowPrefix))) {
                 assert(dictEnd != NULL);
                 if (unlikely(op+length > oend-LASTLITERALS)) {
-                    if (partialDecoding) {
+                    if (unlikely(partialDecoding)) {
                         DEBUGLOG(7, "partialDecoding: dictionary match, close to dstEnd");
                         length = MIN(length, (size_t)(oend-op));
                     } else {
@@ -2097,12 +2219,17 @@ LZ4_decompress_generic(
 
             /* copy match within block */
             cpy = op + length;
+            if (offset >= 32){
+                LZ4_wildCopy64(op, match, cpy);
+                op = cpy;
+                continue;
+            }
 
-            assert((op <= oend) && (oend-op >= 32));
+            assert((op <= oend) && (oend-op >= FASTLOOP_SAFE_DISTANCE));
             if (unlikely(offset<16)) {
                 LZ4_memcpy_using_offset(op, match, cpy, offset);
             } else {
-                LZ4_wildCopy32(op, match, cpy);
+                LZ4_wildCopy64(op, match, cpy);
             }
 
             op = cpy;   /* wildcopy correction */
@@ -2144,9 +2271,9 @@ LZ4_decompress_generic(
                   && (offset >= 8)
                   && (dict==withPrefix64k || match >= lowPrefix) ) {
                     /* Copy the match. */
-                    LZ4_memcpy(op + 0, match + 0, 8);
-                    LZ4_memcpy(op + 8, match + 8, 8);
-                    LZ4_memcpy(op +16, match +16, 2);
+                    KZL_MEMCPY_8(op + 0, match + 0, 8);
+                    KZL_MEMCPY_8(op + 8, match + 8, 8);
+                    KZL_MEMCPY_2(op +16, match +16, 2);
                     op += length + MINMATCH;
                     /* Both stages worked, load the next token. */
                     continue;
@@ -2304,10 +2431,10 @@ LZ4_decompress_generic(
                 op[2] = match[2];
                 op[3] = match[3];
                 match += inc32table[offset];
-                LZ4_memcpy(op+4, match, 4);
+                KZL_MEMCPY_4(op+4, match, 4);
                 match -= dec64table[offset];
             } else {
-                LZ4_memcpy(op, match, 8);
+                KZL_MEMCPY_8(op, match, 8);
                 match += 8;
             }
             op += 8;
@@ -2322,7 +2449,7 @@ LZ4_decompress_generic(
                 }
                 while (op < cpy) { *op++ = *match++; }
             } else {
-                LZ4_memcpy(op, match, 8);
+                KZL_MEMCPY_8(op, match, 8);
                 if (length > 16)  { LZ4_wildCopy8(op+8, match+8, cpy); }
             }
             op = cpy;   /* wildcopy correction */
--- lz4-v1.9.4/lib/Makefile	2022-08-16 06:45:30
+++ lz4-opt/lib/Makefile	2026-05-29 09:20:28
@@ -48,11 +48,11 @@ CPPFLAGS+= $(MOREFLAGS)
 
 CPPFLAGS+= -DXXH_NAMESPACE=LZ4_
 CPPFLAGS+= $(MOREFLAGS)
-CFLAGS  ?= -O3
+CFLAGS  ?= -O3 -march=native
 DEBUGFLAGS:= -Wall -Wextra -Wcast-qual -Wcast-align -Wshadow \
              -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes \
              -Wundef -Wpointer-arith -Wstrict-aliasing=1
-CFLAGS  += $(DEBUGFLAGS)
+CFLAGS  += -flto -fno-plt $(DEBUGFLAGS)
 FLAGS    = $(CFLAGS) $(CPPFLAGS) $(LDFLAGS)
 
 SRCFILES := $(sort $(wildcard *.c))
