diff -uprN lz4_ipp/lz4.c lz4_ipp_new/lz4.c
--- lz4_ipp/lz4.c	2026-03-23 14:47:44.000000000 +0800
+++ lz4_ipp_new/lz4.c	2026-03-27 16:21:07.204000000 +0800
@@ -38,6 +38,10 @@
 #include <ippdc.h>
 #endif
 
+#ifdef __aarch64__
+#include <arm_neon.h>
+#endif
+
 /*-************************************
 *  Tuning parameters
 **************************************/
@@ -328,6 +332,23 @@ typedef enum {
 #define LZ4_memcpy(dst, src, size) memcpy(dst, src, size)
 #endif
 
+#if defined(__GNUC__) && (__GNUC__ >= 4)
+#    define KZL_MEMCPY_2(dst, src, size) __builtin_memcpy(dst, src, size)
+#    define KZL_MEMCPY_4(dst, src, size) __builtin_memcpy(dst, src, size)
+#ifdef __aarch64__
+#    define KZL_MEMCPY_8(dst, src, size)  vst1_u8((dst),  vld1_u8(src))
+#    define KZL_MEMCPY_16(dst, src, size) vst1q_u8((dst), vld1q_u8(src))
+#else
+#    define KZL_MEMCPY_8(dst, src, size)  __builtin_memcpy(dst, src, size)
+#    define KZL_MEMCPY_16(dst, src, size) __builtin_memcpy(dst, src, size)
+#endif
+#else
+#    define KZL_MEMCPY_2(dst, src, size)  memcpy(dst, src, size)
+#    define KZL_MEMCPY_4(dst, src, size)  memcpy(dst, src, size)
+#    define KZL_MEMCPY_8(dst, src, size)  memcpy(dst, src, size)
+#    define KZL_MEMCPY_16(dst, src, size) memcpy(dst, src, size)
+#endif
+
 static unsigned LZ4_isLittleEndian(void)
 {
     const union { U32 u; BYTE c[4]; } one = { 1 };   /* don't use static : performance detrimental */
@@ -417,7 +438,7 @@ void LZ4_wildCopy8(void* dstPtr, const v
     const BYTE* s = (const BYTE*)srcPtr;
     BYTE* const e = (BYTE*)dstEnd;
 
-    do { LZ4_memcpy(d,s,8); d+=8; s+=8; } while (d<e);
+    do { KZL_MEMCPY_8(d,s,8); d+=8; s+=8; } while (d<e);
 }
 
 static const unsigned inc32table[8] = {0, 1, 2,  1,  0,  4, 4, 4};
@@ -450,11 +471,11 @@ LZ4_memcpy_using_offset_base(BYTE* dstPt
         dstPtr[2] = srcPtr[2];
         dstPtr[3] = srcPtr[3];
         srcPtr += inc32table[offset];
-        LZ4_memcpy(dstPtr+4, srcPtr, 4);
+        KZL_MEMCPY_4(dstPtr+4, srcPtr, 4);
         srcPtr -= dec64table[offset];
         dstPtr += 8;
     } else {
-        LZ4_memcpy(dstPtr, srcPtr, 8);
+        KZL_MEMCPY_8(dstPtr, srcPtr, 8);
         dstPtr += 8;
         srcPtr += 8;
     }
@@ -468,11 +489,26 @@ LZ4_memcpy_using_offset_base(BYTE* dstPt
 LZ4_FORCE_INLINE void
 LZ4_wildCopy32(void* dstPtr, const void* srcPtr, void* dstEnd)
 {
+#ifdef __aarch64__
+    int8x16_t v;
+    int8_t *d = (int8_t *)dstPtr;
+    const int8_t *s = (const int8_t *)srcPtr;
+    int8_t *e = (int8_t *)dstEnd;
+    do {
+        v = vld1q_s8(s);
+        vst1q_s8(d, v);
+        v = vld1q_s8(s+16);
+        vst1q_s8(d+16, v);
+        d += 32;
+        s += 32;
+    } while (d < e);
+#else
     BYTE* d = (BYTE*)dstPtr;
     const BYTE* s = (const BYTE*)srcPtr;
     BYTE* const e = (BYTE*)dstEnd;
 
     do { LZ4_memcpy(d,s,16); LZ4_memcpy(d+16,s+16,16); d+=32; s+=32; } while (d<e);
+#endif
 }
 
 /* LZ4_memcpy_using_offset()  presumes :
@@ -490,23 +526,23 @@ LZ4_memcpy_using_offset(BYTE* dstPtr, co
         MEM_INIT(v, *srcPtr, 8);
         break;
     case 2:
-        LZ4_memcpy(v, srcPtr, 2);
-        LZ4_memcpy(&v[2], srcPtr, 2);
-        LZ4_memcpy(&v[4], v, 4);
+        KZL_MEMCPY_2(v, srcPtr, 2);
+        KZL_MEMCPY_2(v+2, srcPtr, 2);
+        KZL_MEMCPY_4(v+4, v, 4);
         break;
     case 4:
-        LZ4_memcpy(v, srcPtr, 4);
-        LZ4_memcpy(&v[4], srcPtr, 4);
+        KZL_MEMCPY_4(v, srcPtr, 4);
+        KZL_MEMCPY_4(v+4, srcPtr, 4);
         break;
     default:
         LZ4_memcpy_using_offset_base(dstPtr, srcPtr, dstEnd, offset);
         return;
     }
 
-    LZ4_memcpy(dstPtr, v, 8);
+    KZL_MEMCPY_8(dstPtr, v, 8);
     dstPtr += 8;
     while (dstPtr < dstEnd) {
-        LZ4_memcpy(dstPtr, v, 8);
+        KZL_MEMCPY_8(dstPtr, v, 8);
         dstPtr += 8;
     }
 }
@@ -723,8 +759,12 @@ LZ4_FORCE_INLINE U32 LZ4_hash5(U64 seque
 
 LZ4_FORCE_INLINE U32 LZ4_hashPosition(const void* const p, tableType_t const tableType)
 {
+#ifndef __aarch64__
     if ((sizeof(reg_t)==8) && (tableType != byU16)) return LZ4_hash5(LZ4_read_ARCH(p), tableType);
     return LZ4_hash4(LZ4_read32(p), tableType);
+#else
+    return LZ4_hash5(LZ4_read_ARCH(p), tableType);
+#endif
 }
 
 LZ4_FORCE_INLINE void LZ4_clearHash(U32 h, void* tableBase, tableType_t const tableType)
@@ -733,9 +773,9 @@ LZ4_FORCE_INLINE void LZ4_clearHash(U32
     {
     default: /* fallthrough */
     case clearedTable: { /* illegal! */ assert(0); return; }
-    case byPtr: { const BYTE** hashTable = (const BYTE**)tableBase; hashTable[h] = NULL; return; }
     case byU32: { U32* hashTable = (U32*) tableBase; hashTable[h] = 0; return; }
     case byU16: { U16* hashTable = (U16*) tableBase; hashTable[h] = 0; return; }
+    case byPtr: { const BYTE** hashTable = (const BYTE**)tableBase; hashTable[h] = NULL; return; }
     }
 }
 
@@ -745,9 +785,9 @@ LZ4_FORCE_INLINE void LZ4_putIndexOnHash
     {
     default: /* fallthrough */
     case clearedTable: /* fallthrough */
-    case byPtr: { /* illegal! */ assert(0); return; }
     case byU32: { U32* hashTable = (U32*) tableBase; hashTable[h] = idx; return; }
     case byU16: { U16* hashTable = (U16*) tableBase; assert(idx < 65536); hashTable[h] = (U16)idx; return; }
+    case byPtr: { /* illegal! */ assert(0); return; }
     }
 }
 
@@ -758,9 +798,9 @@ LZ4_FORCE_INLINE void LZ4_putPositionOnH
     switch (tableType)
     {
     case clearedTable: { /* illegal! */ assert(0); return; }
-    case byPtr: { const BYTE** hashTable = (const BYTE**)tableBase; hashTable[h] = p; return; }
     case byU32: { U32* hashTable = (U32*) tableBase; hashTable[h] = (U32)(p-srcBase); return; }
     case byU16: { U16* hashTable = (U16*) tableBase; hashTable[h] = (U16)(p-srcBase); return; }
+    case byPtr: { const BYTE** hashTable = (const BYTE**)tableBase; hashTable[h] = p; return; }
     }
 }
 
@@ -794,7 +834,7 @@ LZ4_FORCE_INLINE U32 LZ4_getIndexOnHash(
 
 static const BYTE* LZ4_getPositionOnHash(U32 h, const void* tableBase, tableType_t tableType, const BYTE* srcBase)
 {
-    if (tableType == byPtr) { const BYTE* const* hashTable = (const BYTE* const*) tableBase; return hashTable[h]; }
+    if (unlikely(tableType == byPtr)) { const BYTE* const* hashTable = (const BYTE* const*) tableBase; return hashTable[h]; }
     if (tableType == byU32) { const U32* const hashTable = (const U32*) tableBase; return hashTable[h] + srcBase; }
     { const U16* const hashTable = (const U16*) tableBase; return hashTable[h] + srcBase; }   /* default, to ensure a return */
 }
@@ -818,11 +858,10 @@ LZ4_prepareTable(LZ4_stream_t_internal*
      */
     if ((tableType_t)cctx->tableType != clearedTable) {
         assert(inputSize >= 0);
-        if ((tableType_t)cctx->tableType != tableType
+        if (inputSize >= 4 KB || (tableType_t)cctx->tableType != tableType
           || ((tableType == byU16) && cctx->currentOffset + (unsigned)inputSize >= 0xFFFFU)
           || ((tableType == byU32) && cctx->currentOffset > 1 GB)
-          || tableType == byPtr
-          || inputSize >= 4 KB)
+          || tableType == byPtr)
         {
             DEBUGLOG(4, "LZ4_prepareTable: Resetting table in %p", cctx);
             MEM_INIT(cctx->hashTable, 0, LZ4_HASHTABLESIZE);
@@ -973,7 +1012,7 @@ LZ4_FORCE_INLINE int LZ4_compress_generi
     lowLimit = (const BYTE*)source - (dictDirective == withPrefix64k ? dictSize : 0);
 
     /* Update context state */
-    if (dictDirective == usingDictCtx) {
+    if (unlikely(dictDirective == usingDictCtx)) {
         /* Subsequent linked blocks can't use the dictionary. */
         /* Instead, they use the block we just compressed. */
         cctx->dictCtx = NULL;
@@ -997,7 +1036,7 @@ LZ4_FORCE_INLINE int LZ4_compress_generi
         const BYTE* filledIp;
 
         /* Find a match */
-        if (tableType == byPtr) {
+        if (unlikely(tableType == byPtr)) {
             const BYTE* forwardIp = ip;
             int step = 1;
             int searchMatchNb = acceleration << LZ4_skipTrigger;
@@ -1035,7 +1074,7 @@ LZ4_FORCE_INLINE int LZ4_compress_generi
                 if (unlikely(forwardIp > mflimitPlusOne)) goto _last_literals;
                 assert(ip < mflimitPlusOne);
 
-                if (dictDirective == usingDictCtx) {
+                if (unlikely(dictDirective == usingDictCtx)) {
                     if (matchIndex < startIndex) {
                         /* there was no match, try the dictionary */
                         assert(tableType == byU32);
@@ -1047,7 +1086,7 @@ LZ4_FORCE_INLINE int LZ4_compress_generi
                         match = base + matchIndex;
                         lowLimit = (const BYTE*)source;
                     }
-                } else if (dictDirective==usingExtDict) {
+                } else if (unlikely(dictDirective == usingExtDict)) {
                     if (matchIndex < startIndex) {
                         DEBUGLOG(7, "extDict candidate: matchIndex=%5u  <  startIndex=%5u", matchIndex, startIndex);
                         assert(startIndex - matchIndex >= MINMATCH);
@@ -1064,7 +1103,7 @@ LZ4_FORCE_INLINE int LZ4_compress_generi
                 LZ4_putIndexOnHash(current, h, cctx->hashTable, tableType);
 
                 DEBUGLOG(7, "candidate at pos=%u  (offset=%u \n", matchIndex, current - matchIndex);
-                if ((dictIssue == dictSmall) && (matchIndex < prefixIdxLimit)) { continue; }    /* match outside of valid area */
+                if (unlikely(dictIssue == dictSmall) && (matchIndex < prefixIdxLimit)) { continue; }    /* match outside of valid area */
                 assert(matchIndex < current);
                 if ( ((tableType != byU16) || (LZ4_DISTANCE_MAX < LZ4_DISTANCE_ABSOLUTE_MAX))
                   && (matchIndex+LZ4_DISTANCE_MAX < current)) {
@@ -1087,11 +1126,11 @@ LZ4_FORCE_INLINE int LZ4_compress_generi
         /* Encode Literals */
         {   unsigned const litLength = (unsigned)(ip - anchor);
             token = op++;
-            if ((outputDirective == limitedOutput) &&  /* Check output buffer overflow */
+            if (unlikely(outputDirective == limitedOutput) &&  /* Check output buffer overflow */
                 (unlikely(op + litLength + (2 + 1 + LASTLITERALS) + (litLength/255) > olimit)) ) {
                 return 0;   /* cannot compress within `dst` budget. Stored indexes in hash table are nonetheless fine */
             }
-            if ((outputDirective == fillOutput) &&
+            if (unlikely(outputDirective == fillOutput) &&
                 (unlikely(op + (litLength+240)/255 /* litlen */ + litLength /* literals */ + 2 /* offset */ + 1 /* token */ + MFLIMIT - MINMATCH /* min last literals so last match is <= end - MFLIMIT */ > olimit))) {
                 op--;
                 goto _last_literals;
@@ -1120,7 +1159,7 @@ _next_match:
          * - token and *token : position to write 4-bits for match length; higher 4-bits for literal length supposed already written
          */
 
-        if ((outputDirective == fillOutput) &&
+        if (unlikely(outputDirective == fillOutput) &&
             (op + 2 /* offset */ + 1 /* token */ + MFLIMIT - MINMATCH /* min last literals so last match is <= end - MFLIMIT */ > olimit)) {
             /* the match was too close to the end, rewind and go to last literals */
             op = token;
@@ -1141,7 +1180,7 @@ _next_match:
         /* Encode MatchLength */
         {   unsigned matchCode;
 
-            if ( (dictDirective==usingExtDict || dictDirective==usingDictCtx)
+            if ( unlikely(dictDirective==usingExtDict || dictDirective==usingDictCtx)
               && (lowLimit==dictionary) /* match within extDict */ ) {
                 const BYTE* limit = ip + (dictEnd-match);
                 assert(dictEnd > match);
@@ -1160,7 +1199,7 @@ _next_match:
                 DEBUGLOG(6, "             with matchLength=%u", matchCode+MINMATCH);
             }
 
-            if ((outputDirective) &&    /* Check output buffer overflow */
+            if (unlikely(outputDirective) &&    /* Check output buffer overflow */
                 (unlikely(op + (1 + LASTLITERALS) + (matchCode+240)/255 > olimit)) ) {
                 if (outputDirective == fillOutput) {
                     /* Match description too long : reduce it */
@@ -1226,7 +1265,7 @@ _next_match:
             U32 const current = (U32)(ip-base);
             U32 matchIndex = LZ4_getIndexOnHash(h, cctx->hashTable, tableType);
             assert(matchIndex < current);
-            if (dictDirective == usingDictCtx) {
+            if (unlikely(dictDirective == usingDictCtx)) {
                 if (matchIndex < startIndex) {
                     /* there was no match, try the dictionary */
                     matchIndex = LZ4_getIndexOnHash(h, dictCtx->hashTable, byU32);
@@ -1237,7 +1276,7 @@ _next_match:
                     match = base + matchIndex;
                     lowLimit = (const BYTE*)source;  /* required for match length counter */
                 }
-            } else if (dictDirective==usingExtDict) {
+            } else if (unlikely(dictDirective==usingExtDict)) {
                 if (matchIndex < startIndex) {
                     match = dictBase + matchIndex;
                     lowLimit = dictionary;   /* required for match length counter */
@@ -1329,7 +1368,7 @@ LZ4_FORCE_INLINE int LZ4_compress_generi
                 srcSize, dstCapacity);
 
     if ((U32)srcSize > (U32)LZ4_MAX_INPUT_SIZE) { return 0; }  /* Unsupported srcSize, too large (or negative) */
-    if (srcSize == 0) {   /* src == NULL supported if srcSize == 0 */
+    if (unlikely(srcSize == 0)) {   /* src == NULL supported if srcSize == 0 */
         if (outputDirective != notLimited && dstCapacity <= 0) return 0;  /* no output, can't write anything */
         DEBUGLOG(5, "Generating an empty block");
         assert(outputDirective == notLimited || dstCapacity >= 1);
@@ -1789,7 +1828,7 @@ read_variable_length(const BYTE**ip, con
 {
     U32 length = 0;
     U32 s;
-    if (initial_check && unlikely((*ip) >= lencheck)) {    /* overflow detection */
+    if (unlikely((*ip) >= lencheck) && initial_check) {    /* read limit reached */
         *error = initial_error;
         return length;
     }
@@ -1898,7 +1937,7 @@ LZ4_decompress_generic(
 
 	/* Currently the fast loop shows a regression on qualcomm arm chips. */
 #if LZ4_FAST_DEC_LOOP
-        if ((oend - op) < FASTLOOP_SAFE_DISTANCE) {
+        if (unlikely((oend - op) < FASTLOOP_SAFE_DISTANCE)) {
             DEBUGLOG(6, "skip fast decode loop");
             goto safe_decode;
         }
@@ -1914,6 +1953,7 @@ LZ4_decompress_generic(
             assert(!endOnInput || ip <= iend); /* ip < iend before the increment */
 
             /* decode literal length */
+            __builtin_prefetch(ip,0,0);
             if (length == RUN_MASK) {
                 variable_length_error error = ok;
                 length += read_variable_length(&ip, iend-RUN_MASK, (int)endOnInput, (int)endOnInput, &error);
@@ -1940,7 +1980,7 @@ LZ4_decompress_generic(
                     /* We don't need to check oend, since we check it once for each loop below */
                     if (ip > iend-(16 + 1/*max lit + offset + nextToken*/)) { goto safe_literal_copy; }
                     /* Literals can only be 14, but hope compilers optimize if we copy by a register size */
-                    LZ4_memcpy(op, ip, 16);
+                    KZL_MEMCPY_16(op, ip, 16);
                 } else {  /* LZ4_decompress_fast() */
                     /* LZ4_decompress_fast() cannot copy more than 8 bytes at a time :
                      * it doesn't know input length, and relies on end-of-block properties */
@@ -1975,24 +2015,24 @@ LZ4_decompress_generic(
                 }
 
                 /* Fastpath check: Avoids a branch in LZ4_wildCopy32 if true */
-                if ((dict == withPrefix64k) || (match >= lowPrefix)) {
+                if (unlikely((dict == withPrefix64k) || (match >= lowPrefix))) {
                     if (offset >= 8) {
                         assert(match >= lowPrefix);
                         assert(match <= op);
                         assert(op + 18 <= oend);
 
-                        LZ4_memcpy(op, match, 8);
-                        LZ4_memcpy(op+8, match+8, 8);
-                        LZ4_memcpy(op+16, match+16, 2);
+                        KZL_MEMCPY_8(op, match, 8);
+                        KZL_MEMCPY_8(op+8, match+8, 8);
+                        KZL_MEMCPY_2(op+16, match+16, 2);
                         op += length;
                         continue;
             }   }   }
 
             if (checkOffset && (unlikely(match + dictSize < lowPrefix))) { goto _output_error; } /* Error : offset outside buffers */
             /* match starting within external dictionary */
-            if ((dict==usingExtDict) && (match < lowPrefix)) {
+            if (unlikely((dict==usingExtDict) && (match < lowPrefix))) {
                 if (unlikely(op+length > oend-LASTLITERALS)) {
-                    if (partialDecoding) {
+                    if (unlikely(partialDecoding)) {
                         DEBUGLOG(7, "partialDecoding: dictionary match, close to dstEnd");
                         length = MIN(length, (size_t)(oend-op));
                     } else {
@@ -2070,9 +2110,9 @@ LZ4_decompress_generic(
                   && (offset >= 8)
                   && (dict==withPrefix64k || match >= lowPrefix) ) {
                     /* Copy the match. */
-                    LZ4_memcpy(op + 0, match + 0, 8);
-                    LZ4_memcpy(op + 8, match + 8, 8);
-                    LZ4_memcpy(op +16, match +16, 2);
+                    KZL_MEMCPY_8(op + 0, match + 0, 8);
+                    KZL_MEMCPY_8(op + 8, match + 8, 8);
+                    KZL_MEMCPY_2(op + 16, match + 16, 2);
                     op += length + MINMATCH;
                     /* Both stages worked, load the next token. */
                     continue;
@@ -2236,10 +2276,10 @@ LZ4_decompress_generic(
                 op[2] = match[2];
                 op[3] = match[3];
                 match += inc32table[offset];
-                LZ4_memcpy(op+4, match, 4);
+                KZL_MEMCPY_4(op+4, match, 4);
                 match -= dec64table[offset];
             } else {
-                LZ4_memcpy(op, match, 8);
+                KZL_MEMCPY_8(op, match, 8);
                 match += 8;
             }
             op += 8;
@@ -2254,7 +2294,7 @@ LZ4_decompress_generic(
                 }
                 while (op < cpy) { *op++ = *match++; }
             } else {
-                LZ4_memcpy(op, match, 8);
+                KZL_MEMCPY_8(op, match, 8);
                 if (length > 16)  { LZ4_wildCopy8(op+8, match+8, cpy); }
             }
             op = cpy;   /* wildcopy correction */
